File: java/exercises/04-expert/ex094_event_bus/EventBus.java
package fewolearning.exercises.expert.ex094_event_bus;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

/*
Exercise 094 - Event bus (expert).

Goal:   Register subscribers per event type and dispatch events to them synchronously.
Drills: publish/subscribe, synchronous dispatch.
*/
public final class EventBus {
    private final Map<Class<?>, List<Consumer<Object>>> subscribers = new HashMap<>();

    public <T> void subscribe(Class<T> eventType, Consumer<T> handler) {
        throw new UnsupportedOperationException("TODO");
    }

    public void publish(Object event) {
        throw new UnsupportedOperationException("TODO");
    }
}

File: kotlin/exercises/02-intermediate/ex067_shared_flow_events/EventBus.kt
package fewolearning.exercises.intermediate.ex067_shared_flow_events

import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow

/*
Exercise 067 - SharedFlow events (intermediate).

Goal:   Broadcast one-off events to all current subscribers via a SharedFlow.
Drills: SharedFlow, broadcast events.
*/
class EventBus {
    private val _events = MutableSharedFlow<String>()
    val events: SharedFlow<String> get() = _events

    suspend fun publish(event: String) {
        TODO()
    }
}
